SmartDeviceLink/SDLSystemCapabilityManager.m

```diff
@@ -119,43 +149,128 @@ - (void)sdl_displayLayoutResponse:(SDLRPCResponseNotification *)notification {
+    [self.connectionManager sendConnectionRequest:getSystemCapability withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        if (error != nil) {
+            // An error is returned if the request was unsuccessful or if a Generic Response was returned
+            if (self.systemCapabilityHandler == nil) { return; }
+            return self.systemCapabilityHandler(error, self);
+        }
+
+        SDLGetSystemCapabilityResponse *getSystemCapabilityResponse = (SDLGetSystemCapabilityResponse *)response;
+        if (!getSystemCapabilityResponse.resultCode.boolValue) { return; }
+        [self sdl_saveSystemCapability:getSystemCapabilityResponse.systemCapability];
+    }];
+}
+
+/**
+ *  Saves a system capability. All system capabilities will update with the full object except for app services. For app services only the updated app service capabilities will be included in the `SystemCapability` sent from Core. The cached `appServicesCapabilities` will be updated with the new `appService` data.
+ *
+ *  @param systemCapability The system capability
+ */
+- (void)sdl_saveSystemCapability:(SDLSystemCapability *)systemCapability {
+    SDLSystemCapabilityType systemCapabilityType = systemCapability.systemCapabilityType;
 
     if ([systemCapabilityType isEqualToEnum:SDLSystemCapabilityTypePhoneCall]) {
-        self.phoneCapability = systemCapabilityResponse.phoneCapability;
+        self.phoneCapability = systemCapability.phoneCapability;
     } else if ([systemCapabilityType isEqualToEnum:SDLSystemCapabilityTypeNavigation]) {
-        self.navigationCapability = systemCapabilityResponse.navigationCapability;
+        self.navigationCapability = systemCapability.navigationCapability;
     } else if ([systemCapabilityType isEqualToEnum:SDLSystemCapabilityTypeRemoteControl]) {
-        self.remoteControlCapability = systemCapabilityResponse.remoteControlCapability;
+        self.remoteControlCapability = systemCapability.remoteControlCapability;
     } else if ([systemCapabilityType isEqualToEnum:SDLSystemCapabilityTypeVideoStreaming]) {
-        self.videoStreamingCapability = systemCapabilityResponse.videoStreamingCapability;
+        self.videoStreamingCapability = systemCapability.videoStreamingCapability;
     } else if ([systemCapabilityType isEqualToEnum:SDLSystemCapabilityTypeAppServices]) {
-        self.appServicesCapabilities = systemCapabilityResponse.appServicesCapabilities;
+        if (!self.appServicesCapabilities) {
+            self.appServicesCapabilities = systemCapability.appServicesCapabilities;
+        } else {
+            NSMutableDictionary *cachedAppServicesCapabilities = [NSMutableDictionary dictionary];
+            for (NSUInteger i = 0; i < self.appServicesCapabilities.appServices.count; i += 1) {
+                SDLAppServiceRecord *record = self.appServicesCapabilities.appServices[i].updatedAppServiceRecord;
```

Comment:
I don't think this is probably the best approach. It's going to be super slow. It's probably better to just store a mutable dictionary of the values privately and then the public `appServicesCapability` property has a getter that assembles it based the `[dictionary allValues]` call. Hopefully that makes sense.